### DISCUSSION TITLE
Next + Typescript production build looks bloated with typescript files

### DISCUSSION THREAD

**Discussion by skasyn** [vercel/next.js#17746] [Help]
Hello ! I just started to work on a company that uses Next js (🙌). 

The company is almost full of backend dev with little to no experience on frontend. They are using hygen, a codegen tool, to build UI. The tool was setup by a freelance developer that left 1 month ago so no one can answer my question right now.

The project is somewhat huge, with more than 10 000 ts/tsx files. The website still feels fast but the build takes ages. Out of curiosity, I ran the nextjs bundle analyzer and the result shocked me :

Client : 
![image](https://user-images.githubusercontent.com/38563623/95583069-e2d6ac80-0a3b-11eb-87c9-4644084156fd.png)

Server :
![image](https://user-images.githubusercontent.com/38563623/95583104-ef5b0500-0a3b-11eb-9c2e-19365800bb27.png)

Maybe I find it surprising because I lack experience, but those numbers still looks big. What I find surprising is this : 
![image](https://user-images.githubusercontent.com/38563623/95583361-4a8cf780-0a3c-11eb-8a03-183f90edf17c.png)
(client side)

This file is generated and contains graphql types for typing queries. I feel like I'm missing something : shouldnt typescript files be converted to .js and types **simply removed from production build** ? What is the point of keeping types in production ?

What is even weirder is the server side : 
![image](https://user-images.githubusercontent.com/38563623/95583619-b40d0600-0a3c-11eb-89ed-3cc41d62d782.png)
This is a small sample of the output but it shows what bugs me : why is the huge typescript type file duplicated and bundled with almost every pages of the application ? No wonder why the server is 230 Mb.

Maybe it's because I simply don't understand how bundling works and how to read the output of next bundle analyzer but I can't find any ressources about this and as I said, no one Is able to explain it to me in the office. I've looked at the documentation without finding any reference about this.

The next.config.js seems simple enough, I don't think it has anything to do with this : 
```js
module.exports = {
  // target: 'serverless',
  webpackDevMiddleware: (config) => {
    // eslint-disable-next-line no-param-reassign
    config.watchOptions = {
      poll: 1000,
      aggregateTimeout: 300,
    };

    return config;
  },
  sassOptions: {
    includePaths: ['./styles'],
  },
};
```

On the other hand, there is a .babelrc file but since I have little to no experience with babel, maybe it lacks a module that would "clean" the build ?  :
```js
{
  "presets": [
    "next/babel"
  ],
  "plugins": [
    "import-graphql",
    "graphql-tag",
    [
      "module-resolver",
      {
        "alias": {
          "@resolvers": "./resolvers",
          "@components": "./components",
          "@containers": "./containers",
          "@pages": "./pages",
          "@schema": "./schema",
          "@utils": "./utils",
          "@translations": "./translations",
          "@styles": "./styles"
        }
      }
    ],
    [
      "react-intl"
    ]
  ]
}
```

I don't even know if this is a bug or even a problem. All I know is that I've never seen that and that it feels wrong. If someone could give me some insight about this, I'll be grateful.

TL:DR 

Next.js production build is full of one typescript file full of types that is duplicated more than 200 times in the server build.

- is it normal to see ts files in production build ?
- is it normal to see a >200 Mb production server filled with the same 1.Mb ts file duplicated ?
- where can I find ressource about next js builds

**Comment by sct** [comment]
I am having the same issue. My docker builds are well over 300MB after pruning all dev dependencies due to next's dependencies that it's forcefully including. There must be a way to strip this stuff?
